File: src/dotnet/TraceMap.Core/FactFactory.cs
using System.Security.Cryptography;
using System.Text;

namespace TraceMap.Core;

public static class FactFactory
{
    public static CodeFact Create(
        ScanManifest manifest,
        string factType,
        string ruleId,
        string evidenceTier,
        EvidenceSpan evidence,
        string? projectPath = null,
        string? sourceSymbol = null,
        string? targetSymbol = null,
        string? contractElement = null,
        IReadOnlyDictionary<string, string>? properties = null)
    {
        var stableProperties = new SortedDictionary<string, string>(StringComparer.Ordinal);
        foreach (var (key, value) in properties ?? new Dictionary<string, string>())
        {
            stableProperties[key] = value;
        }
        var factId = CreateFactId(
            manifest.ScanId,
            factType,
            ruleId,
            evidence.FilePath,
            evidence.StartLine,
            evidence.EndLine,
            projectPath,
            sourceSymbol,
            targetSymbol,
            contractElement,
            stableProperties);

        return new CodeFact(
            factId,
            manifest.ScanId,
            manifest.RepoName,
            manifest.CommitSha,
            projectPath,
            factType,
            ruleId,
            evidenceTier,
            sourceSymbol,
            targetSymbol,
            contractElement,
            evidence,
            stableProperties);
    }

    private static string CreateFactId(
        string scanId,
        string factType,
        string ruleId,
        string filePath,
        int startLine,
        int endLine,
        string? projectPath,
        string? sourceSymbol,
        string? targetSymbol,
        string? contractElement,
        IReadOnlyDictionary<string, string> properties)
    {
        var builder = new StringBuilder();
        builder.Append(scanId).Append('|')
            .Append(factType).Append('|')
            .Append(ruleId).Append('|')
            .Append(filePath).Append('|')
            .Append(startLine).Append('|')
            .Append(endLine).Append('|')
            .Append(projectPath).Append('|')
            .Append(sourceSymbol).Append('|')
            .Append(targetSymbol).Append('|')
            .Append(contractElement);

        foreach (var (key, value) in properties)
        {
            builder.Append('|').Append(key).Append('=').Append(value);
        }

        return "fact-" + Hash(builder.ToString(), 20);
    }

    public static string Hash(string value, int length)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(value));
        var hex = Convert.ToHexString(bytes).ToLowerInvariant();
        return hex[..Math.Min(length, hex.Length)];
    }
}

File: src/jvm/src/main/java/com/tracemap/jvm/facts/FactFactory.java
package com.tracemap.jvm.facts;

import com.tracemap.jvm.model.CodeFact;
import com.tracemap.jvm.model.EvidenceSpan;
import com.tracemap.jvm.model.ScanManifest;
import com.tracemap.jvm.util.Hashes;
import java.util.Map;
import java.util.TreeMap;

public final class FactFactory {
    private FactFactory() {
    }

    public static EvidenceSpan evidence(String filePath, int startLine, int endLine, String extractorId, String extractorVersion) {
        int safeStart = Math.max(1, startLine);
        int safeEnd = Math.max(safeStart, endLine);
        return new EvidenceSpan(filePath, safeStart, safeEnd, null, extractorId, extractorVersion);
    }

    public static EvidenceSpan evidence(String filePath, int startLine, int endLine, String snippetHash, String extractorId, String extractorVersion) {
        int safeStart = Math.max(1, startLine);
        int safeEnd = Math.max(safeStart, endLine);
        return new EvidenceSpan(filePath, safeStart, safeEnd, snippetHash, extractorId, extractorVersion);
    }

    public static CodeFact create(
        ScanManifest manifest,
        String factType,
        String ruleId,
        String evidenceTier,
        EvidenceSpan evidence,
        String projectPath,
        String sourceSymbol,
        String targetSymbol,
        String contractElement,
        Map<String, String> properties) {
        TreeMap<String, String> stableProperties = new TreeMap<>();
        if (properties != null) {
            for (Map.Entry<String, String> entry : properties.entrySet()) {
                if (entry.getValue() != null) {
                    stableProperties.put(entry.getKey(), entry.getValue());
                }
            }
        }
        String factId = createFactId(
            manifest.scanId(),
            factType,
            ruleId,
            evidence.filePath(),
            evidence.startLine(),
            evidence.endLine(),
            projectPath,
            sourceSymbol,
            targetSymbol,
            contractElement,
            stableProperties);
        return new CodeFact(
            factId,
            manifest.scanId(),
            manifest.repoName(),
            manifest.commitSha(),
            projectPath,
            factType,
            ruleId,
            evidenceTier,
            sourceSymbol,
            targetSymbol,
            contractElement,
            evidence,
            Map.copyOf(stableProperties));
    }

    public static String createFactId(
        String scanId,
        String factType,
        String ruleId,
        String filePath,
        int startLine,
        int endLine,
        String projectPath,
        String sourceSymbol,
        String targetSymbol,
        String contractElement,
        Map<String, String> properties) {
        StringBuilder builder = new StringBuilder();
        builder.append(scanId).append('|')
            .append(factType).append('|')
            .append(ruleId).append('|')
            .append(filePath).append('|')
            .append(startLine).append('|')
            .append(endLine).append('|')
            .append(projectPath == null ? "" : projectPath).append('|')
            .append(sourceSymbol == null ? "" : sourceSymbol).append('|')
            .append(targetSymbol == null ? "" : targetSymbol).append('|')
            .append(contractElement == null ? "" : contractElement);
        properties.entrySet().stream()
            .sorted(Map.Entry.comparingByKey())
            .forEach(entry -> builder.append('|').append(entry.getKey()).append('=').append(entry.getValue()));
        return "fact-" + Hashes.sha256(builder.toString(), 20);
    }
}
